app/controllers/concerns/foreman/controller/puppet/hosts_controller_extensions.rb

```diff
@@ -0,0 +1,164 @@
+module Foreman::Controller::Puppet::HostsControllerExtensions
+  extend ActiveSupport::Concern
```

Comment:
Makes sense. Thanks. I agree it would need first for the UI to be rethinked/redesigned for the separate controller to make sense.